### PR TITLE
Wrap around Rucio write APIs for data registration

### LINKED ISSUE #9068: Extend Rucio Service class with data injection/subscription API
Once we have basic read Rucio operations like:
* checking whether a dataset exist in Rucio
* data discovery/location in Rucio

we can then start writing the WMCore APIs for registering CMS files/blocks/dataset and creating subscriptions (rules).

Related to: https://its.cern.ch/jira/browse/CMSRUCIO-97
and 
https://its.cern.ch/jira/browse/CMSRUCIO-37

Reference of writing operations:
https://github.com/dmwm/CMSKubernetes/blob/master/docker/rucio-sync/scripts/BlockSyncer.py#L78

### PR DESCRIPTION
**Pull request by amaltaro**
Fixes #9068 

#### Status
In development

#### Description
Will be filled later

#### Is it backward compatible (if not, which system it affects?)
yes

#### Related PRs
See further examples and usage in the CMSRucio module developed by the CRAB team: https://github.com/dmwm/CRABServer/pull/5954

#### External dependencies / deployment changes
none

### CODE REVIEW

**Comment by amaltaro** [comment]
I have done further changes to this patch and managed to get some things tested, it's looking good from my side.

Once I get further rights in Rucio/RSE, I still need to test adding replicas, closing a block and creating a basic rule.

It's likely in a good shape for a code review now, Eric and Todor.

**Comment by amaltaro** [comment]
Only missing bit is now creating a Rucio rule. Eric, would you suggest some specific test (that won't cause any problems) for a rule for:
```
contName = '/RelValProdMinBias/DMWM_Test-ProdMinBias_TaskChain_ProdMinBias_Feb2020_Val_AlanPrivv2-v11/GEN-SIM'
scope = 'user.amaltaro'
```

in the `int` instance.

**Comment by ericvaandering** [comment]
You can put a rule on where you just placed it. Or you can also put a rule to move it to T1_US_FNAL_Disk_Test.

I'll be curious to see what happens with the user.amaltaro scope. Hopefully @dciangot planned for that and our LFN to PFN conversion works for it.

File: src/python/WMCore/Services/Rucio/Rucio.py [hunk 1]
```diff
@@ -8,11 +8,10 @@
 from __future__ import division, print_function, absolute_import
 
 import logging
-
+from copy import deepcopy
 from rucio.client import Client
 from rucio.common.exception import AccountNotFound, DataIdentifierNotFound,\
```
Comment (ericvaandering) [inline src/python/WMCore/Services/Rucio/Rucio.py]: Prettier way to do this is use () around the exception names. Preferred now too.

File: src/python/WMCore/Services/Rucio/Rucio.py [hunk 2]
```diff
@@ -229,3 +239,246 @@ def getPFN(self, nodes=None, lfns=None, destination=None, protocol='srmv2', cust
...
+        except DataIdentifierAlreadyExists:
+            self.logger.debug("Block name already exists in Rucio: %s", name)
+            response = True
+        except Exception as ex:
+            self.logger.error("Exception creating block: %s. Error: %s", name, str(ex))
+
+        # then attach this block recently created to its container
+        if response and attach:
+            container = name.split('#')[0]
+            response = self.attachDIDs(kwargs.get('rse'), container, name, scope)
```
Comment (ericvaandering) [inline src/python/WMCore/Services/Rucio/Rucio.py]: You can get an error here if the block already existed and was already attached to the container (I think).

Comment (ericvaandering) [reply src/python/WMCore/Services/Rucio/Rucio.py]: Yeah, you catch a similar error in the next section.

File: src/python/WMCore/Services/Rucio/Rucio.py [hunk 3]
```diff
@@ -229,3 +239,246 @@ def getPFN(self, nodes=None, lfns=None, destination=None, protocol='srmv2', cust
...
+
+        # TODO: test to make sure 'state' is a valid argument
+        response = False
+        try:
+            # add_replicas(rse, files, ignore_availability=True)
+            response = self.cli.add_replicas(rse, files, ignoreAvailability)
+        except Exception as ex:
+            self.logger.error("Failed to add replicas for: %s and block: %s. Error: %s", files, block, str(ex))
+
+        if response:
```
Comment (ericvaandering) [inline src/python/WMCore/Services/Rucio/Rucio.py]: I'm not sure the logic here makes sense to try to attach DIDs every time a replica is made. 

In general, the way I'd think about it is 
- tell Rucio which DIDs exist and how they are related
- tell Rucio where the replicas of files are

So attaching files to block is part of the first part.

File: src/python/WMCore/Services/Rucio/Rucio.py [hunk 4]
```diff
@@ -229,3 +239,246 @@ def getPFN(self, nodes=None, lfns=None, destination=None, protocol='srmv2', cust
...
+        kwargs.setdefault('account', self.rucioParams.get('account'))
+        kwargs.setdefault('locked', False)
+        kwargs.setdefault('notify', 'N')
+        kwargs.setdefault('purge_replicas', False)
+        kwargs.setdefault('ignore_availability', False)
+        kwargs.setdefault('ask_approval', False)
+        kwargs.setdefault('asynchronous', False)
+        kwargs.setdefault('priority', 3)
+
+        if isinstance(names, basestring):
```
Comment (ericvaandering) [inline src/python/WMCore/Services/Rucio/Rucio.py]: This is not going to work in python3. Is there another way?

**Review by ericvaandering — CHANGES_REQUESTED** [review]
In general it looks good. My only substantive comment in on the logic of did attachment.

File: src/python/WMCore/Services/Rucio/Rucio.py (hunk 1 above)
Comment (amaltaro) [reply src/python/WMCore/Services/Rucio/Rucio.py]: will update it

File: src/python/WMCore/Services/Rucio/Rucio.py (hunk 2 above)
Comment (amaltaro) [reply src/python/WMCore/Services/Rucio/Rucio.py]: Yes, the error handling is done on the `attachDIDs` method

File: src/python/WMCore/Services/Rucio/Rucio.py (hunk 3 above)
Comment (amaltaro) [reply src/python/WMCore/Services/Rucio/Rucio.py]: Ok, this might be where I get confused.
When we add a new replica (which is a file at a given location), doesn't it meet your second bullet comment?

In short, we add a replica also providing their RSE availability, then we attach those replicas to a block.

File: src/python/WMCore/Services/Rucio/Rucio.py (hunk 4 above)
Comment (amaltaro) [reply src/python/WMCore/Services/Rucio/Rucio.py]: I will check if it's not a list instead ;)

**Comment by amaltaro** [comment]
Thanks, Eric. I applied changes as you suggested, the only bit still opened is related to the replica logic (I replied to your comment in there).

**Comment by ericvaandering** [comment]
Right, but replicas are not attached to blocks, LFNs are.

Sent from a mobile device.

On Feb 7, 2020, at 11:08, Alan Malta Rodrigues <notifications@github.com> wrote:

﻿

@amaltaro commented on this pull request.

________________________________

In src/python/WMCore/Services/Rucio/Rucio.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dmwm_WMCore_pull_9401-23discussion-5Fr376504661&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=EHaoB-POFWGrYFvPXoj1bQ&m=QZXWFsYbwULEnLFShHYAiVXcV5V0ZjM4G9rWaadWiCQ&s=BXnOij-q0gD1Rq2OEmBMclMy9AtbTirL2DwDA_iOQFY&e=>:

> +        :return: a boolean to represent whether it succeeded or not
+        """
+        if isinstance(files, dict):
+            files = [files]
+        for item in files:
+            item['scope'] = scope
+
+        # TODO: test to make sure 'state' is a valid argument
+        response = False
+        try:
+            # add_replicas(rse, files, ignore_availability=True)
+            response = self.cli.add_replicas(rse, files, ignoreAvailability)
+        except Exception as ex:
+            self.logger.error("Failed to add replicas for: %s and block: %s. Error: %s", files, block, str(ex))
+
+        if response:


Ok, this might be where I get confused.
When we add a new replica (which is a file at a given location), doesn't it meet your second bullet comment?

In short, we add a replica also providing their RSE availability, then we attach those replicas to a block.

**Comment by amaltaro** [comment]
Within the WM system, I don't think we will ever have a LFN that does not have a location (besides some non-EDM T0 files). That's why when we create a replica through `createReplica`, we also pass those LFNs to be attached to a block (by the end of the same method).

Are you suggesting to have another method simply to add a new did (a LFN in this case)?

BTW, the more I read this code the more I think I should rename those methods from `createxxx` to `addxxx`. Will do so soon.

**Comment by amaltaro** [comment]
@dciangot Diego, you might want to have a look/review this PR as well. I'm planning to get it in today/tomorrow.

**Comment by amaltaro** [comment]
Now that I managed to test the rule creation (and fixed a problem there), I decided to implement a `getRule` API as well.
Things have been tested and seem to be working fine.

@ericvaandering could you please clarify my previous comment? AFAICS there is nothing else to be fixed, but please let me know if I'm wrong.

**Comment by ericvaandering** [comment]
One could imagine a mode in which you are creating the structure of the files before you are ready to create the replicas. 

It probably doesn't matter on a technical level, it's just a question of how you think about the objects and their relations. And separating file creation and attachment from replica creation is how Rucio thinks about these concepts.

**Comment by amaltaro** [comment]
Yes, it indeed makes sense for the data management tool. 
We might revisit it in the future once all this becomes more mature in the WM context.

Thanks for looking into it, Eric.

**Comment by amaltaro** [comment]
FYI @belforte . It will be in `1.3.1.pre1` wmcore tag.

**Comment by belforte** [comment]
Thanks Alan. @dciangot please take note.

**Comment by amaltaro** [comment]
@todor-ivanov can you please merge this PR and cut a new tag `1.3.1.pre1`? Thanks

**Review by todor-ivanov — APPROVED** [review]
good to go
